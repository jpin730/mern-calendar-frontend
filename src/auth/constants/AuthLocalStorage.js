export const AUTH_LOCAL_STORAGE = {
  token: "token",
  tokenInitDate: "token-init-date",
};
