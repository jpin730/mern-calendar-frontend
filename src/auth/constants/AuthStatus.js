export const AUTH_STATUS = {
  authenticated: "authenticated",
  notAuthenticated: "not-authenticated",
};
