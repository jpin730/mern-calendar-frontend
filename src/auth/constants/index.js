export * from "./AuthLocalStorage";
export * from "./AuthStatus";
