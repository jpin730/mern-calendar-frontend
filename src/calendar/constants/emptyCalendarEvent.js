import { addHours } from "date-fns";

const todayDate = new Date();
todayDate.setHours(todayDate.getHours() + 1, 0, 0, 0);

export const emptyCalendarEvent = {
  title: "",
  notes: "",
  start: todayDate,
  end: addHours(todayDate, 2),
};
