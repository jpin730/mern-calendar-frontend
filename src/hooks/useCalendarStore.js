import { useDispatch, useSelector } from "react-redux";

import {
  onAddNewEvent,
  onDeleteEvent,
  onSetActiveEvent,
  onUpdateEvent,
} from "../store/calendar/calendarSlice";
import { calendarApi } from "../api";

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  const { events, activeEvent } = useSelector((state) => state.calendar);

  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  const startSavingEvent = async (calendarEvent) => {
    if (calendarEvent._id) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      const { data } = await calendarApi.post("/events", calendarEvent);
      dispatch(onAddNewEvent({ ...calendarEvent, id: data.createdEvent.id }));
    }
  };

  const startDeletingEvent = () => {
    dispatch(onDeleteEvent());
  };

  return {
    activeEvent,
    events,
    hasEventSelected: !!activeEvent,
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
  };
};
