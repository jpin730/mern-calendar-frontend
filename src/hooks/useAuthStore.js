import { useDispatch, useSelector } from "react-redux";

import { clearErrorMessage, onChecking, onLogin, onLogout } from "../store";
import { AUTH_LOCAL_STORAGE } from "../auth";
import { calendarApi } from "../api";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/login", {
        email,
        password,
      });
      localStorage.setItem(AUTH_LOCAL_STORAGE.token, data.token);
      localStorage.setItem(
        AUTH_LOCAL_STORAGE.tokenInitDate,
        new Date().getTime
      );
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch {
      dispatch(onLogout("Wrong credentials"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      });
    }
  };

  const startRegister = async ({ name, email, password }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/user", {
        name,
        email,
        password,
      });
      localStorage.setItem(AUTH_LOCAL_STORAGE.token, data.token);
      localStorage.setItem(
        AUTH_LOCAL_STORAGE.tokenInitDate,
        new Date().getTime
      );
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      dispatch(onLogout(error?.response?.data?.msg || "Registration error"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      });
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem(AUTH_LOCAL_STORAGE.token);
    if (!token) {
      return dispatch(onLogout());
    }

    try {
      const { data } = await calendarApi.get("auth/token");
      localStorage.setItem(AUTH_LOCAL_STORAGE.token, data.token);
      localStorage.setItem(
        AUTH_LOCAL_STORAGE.tokenInitDate,
        new Date().getTime()
      );
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      localStorage.clear();
      dispatch(onLogout());
    }
  };

  return {
    status,
    user,
    errorMessage,
    startLogin,
    startRegister,
    checkAuthToken,
  };
};
