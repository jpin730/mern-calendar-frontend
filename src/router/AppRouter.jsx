import { Navigate, Route, Routes } from "react-router-dom";

import { AUTH_STATUS, LoginPage } from "../auth";
import { CalendarPage } from "../calendar";

export const AppRouter = () => {
  const authStatus = AUTH_STATUS.notAuthenticated;
  return (
    <Routes>
      {authStatus === AUTH_STATUS.notAuthenticated ? (
        <Route path="/auth/*" element={<LoginPage />} />
      ) : (
        <Route path="/*" element={<CalendarPage />} />
      )}

      <Route path="/*" element={<Navigate to="/auth/login" />} />
    </Routes>
  );
};
