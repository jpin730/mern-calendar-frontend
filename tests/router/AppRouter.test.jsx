import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { AUTH_STATUS } from "../../src/auth/constants/AuthStatus";
import { useAuthStore } from "../../src/hooks/useAuthStore";
import { AppRouter } from "../../src/router/AppRouter";

jest.mock("../../src/hooks/useAuthStore");

jest.mock("../../src/calendar", () => ({
  CalendarPage: () => <h1>CalendarPage</h1>,
}));

describe("AppRouter", () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test("should show loading and call checkAuthToken", () => {
    useAuthStore.mockReturnValue({
      status: AUTH_STATUS.checking,
      checkAuthToken: mockCheckAuthToken,
    });

    render(<AppRouter />);

    expect(screen.getByRole("heading", { level: 3 }).innerHTML).toBe(
      "Loading..."
    );
    expect(mockCheckAuthToken).toHaveBeenCalledWith();
  });

  test("should show login page if user is not authenticated", () => {
    useAuthStore.mockReturnValue({
      status: AUTH_STATUS.notAuthenticated,
      checkAuthToken: mockCheckAuthToken,
    });

    const { container } = render(
      <MemoryRouter initialEntries={["/home"]}>
        <AppRouter />
      </MemoryRouter>
    );

    const h3Elements = screen.getAllByRole("heading", { level: 3 });

    expect(h3Elements.at(0).innerHTML).toBe("Login");
    expect(h3Elements.at(1).innerHTML).toBe("Create an account");
    expect(container).toMatchSnapshot();
  });

  test("should show login page if user is not authenticated", () => {
    useAuthStore.mockReturnValue({
      status: AUTH_STATUS.authenticated,
      checkAuthToken: mockCheckAuthToken,
    });

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    expect(screen.getByRole("heading", { level: 1 }).innerHTML).toBe(
      "CalendarPage"
    );
  });
});
