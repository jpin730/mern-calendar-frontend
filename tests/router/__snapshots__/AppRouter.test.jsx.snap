// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRouter should show login page if user is not authenticated 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="min-vh-100 d-flex align-items-center justify-content-center"
    >
      <div
        class="row w-100 border rounded shadow bg-white py-5 px-3 g-5"
      >
        <div
          class="col-lg-6 m-0"
        >
          <h3>
            Login
          </h3>
          <hr />
          <form>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="loginEmail"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="loginPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="form-group mb-3 text-end"
            >
              <input
                class="btn btn-primary px-5"
                type="submit"
                value="Login"
              />
            </div>
          </form>
        </div>
        <div
          class="col-lg-6 m-0"
        >
          <h3>
            Create an account
          </h3>
          <hr />
          <form>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="registerName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="registerEmail"
                placeholder="Email"
                type="email"
                value=""
              />
            </div>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="registerPassword"
                placeholder="Password"
                type="password"
                value=""
              />
            </div>
            <div
              class="form-group mb-3"
            >
              <input
                class="form-control"
                name="registerPasswordConfirm"
                placeholder="Confirm password"
                type="password"
                value=""
              />
            </div>
            <div
              class="form-group mb-3 text-end"
            >
              <input
                class="btn btn-primary px-5"
                type="submit"
                value="Register"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
